skxray/dpc.py

```diff
@@ -1,276 +1,462 @@
+        
+    gx_padding = np.zeros((pad_row, pad_col), dtype='d')
+    gy_padding = np.zeros((pad_row, pad_col), dtype='d')
+    
+    roi_slice = (slice(padding * rows, (padding + 1) * rows),
+                 slice(padding * cols, (padding + 1) * cols))
+    gx_padding[roi_slice] = gx
+    gy_padding[roi_slice] = gy
+
+    tx = np.fft.fftshift(np.fft.fft2(gx_padding))
+    ty = np.fft.fftshift(np.fft.fft2(gy_padding))
+
+    mid_col = pad_col // 2 + 1
+    mid_row = pad_row // 2 + 1
+    ax = (2 * np.pi * np.arange(1 - mid_col, pad_col - mid_col + 1) / 
+         (pad_col * scan_xstep))
+    ay = (2 * np.pi * np.arange(1 - mid_row, pad_row - mid_row + 1) / 
+         (pad_row * scan_ystep))
+
+    kappax, kappay = np.meshgrid(ax, ay)
+    div_v = kappax ** 2 * (1 - weighting) + kappay ** 2 * weighting
+
+    c = -1j * (kappax * tx * (1 - weighting) + kappay * ty * weighting) / div_v
+    c = np.fft.ifftshift(np.where(div_v==0, 0, c))
+
+    phase = np.fft.ifft2(c)[roi_slice].real
+
+    return phase
+
+
+def dpc_runner(ref, image_sequence, start_point, pixel_size, focus_to_det, 
+               scan_rows, scan_cols, scan_xstep, scan_ystep, energy, padding=0, 
+               weighting=0.5, solver='Nelder-Mead', roi=None, bad_pixels=None, 
+               negate=True, scale=True):
+    """
+    Controller function to run the whole Differential Phase Contrast (DPC) 
+    imaging calculation.
+    
+    Parameters
+    ----------
+    ref : ndarray
+        The reference image for a DPC calculation.
+
+    image_sequence : iterable of 2D arrays
+        Return diffraction patterns (2D Numpy arrays) when iterated over.
+        
+    start_point : list
+        start_point[0], start-searching value for the amplitude of the sample 
+        transmission function at one scanning point.
+        start_point[1], start-searching value for the phase gradient (along x 
+        or y direction) of the sample transmission function at one scanning 
+        point.
+
+    pixel_size : tuple
+        Physical pixel (a rectangle) size of the detector in um.
+
+    focus_to_det : float
+        Focus to detector distance in um.
+
+    scan_rows : int
+        Number of scanned rows.
+
+    scan_cols : int
+        Number of scanned columns.
+
+    scan_xstep : float
+        Scanning step size in x direction (in micro-meter).
+
+    scan_ystep : float
+        Scanning step size in y direction (in micro-meter).
+
+    energy : float
+        Energy of the scanning x-ray in keV.
+    
+    padding : int, optional
+        Pad a N-by-M array to be a (N*(2*padding+1))-by-(M*(2*padding+1)) array 
+        with the image in the middle with a (N*padding, M*padding) thick edge 
+        of zeros. Default is 0.
+        padding = 0 --> v (the original image, size = (N, M))
+                        0 0 0
+        padding = 1 --> 0 v 0 (the padded image, size = (3 * N, 3 * M))
+                        0 0 0
+
+    weighting : float, valid in [0, 1], optional
```

Comment:
Don't include the range in the type spec, it will break the docstring parsing to wrap this for vistrails